### PR TITLE
756 ABM benchmark

### LINKED ISSUE #756: Benchmark for ABM
### Feature description

We should make sure performance of the ABM doesn't deteriorate. Therefore we should have benchmarks in the CI. We would need a dedicated runner, e.g. on our HPDA Cluster that allows consistent benchmarks. The CI check is passed if runtime is within measurement tolerance of a previous run. To get the tolerance right will probably produce some false positives in the beginning.

If the runtime is measurably greater, there should be additional discussion whether there are any performance bugs (e.g. unnecessary copies) or the feature can be implemented more efficiently. Also maybe the feature isn't worth it or needs to be optional if it is too slow.

Faster runtime could indicate a bug. But maybe our test coverage is good enough that this additional check is likely false positive?

When this works for the ABM we can introduce it for the other models as well. But performance is more important and also more likely to deteriorate in the ABM.

### Additional context

_No response_

### Checklist

- [X] Attached labels, especially loc:: or model:: labels.
- [X] Linked to project.

### PR DESCRIPTION
**Pull request by dabele**
# Changes and Information

Added a benchmark for the ABM, currently to be run manually, will be added to CI in the future.

Please confirm that the benchmark covers all features, locations, infections, etc, I may have missed some. 

Closes #756

## Merge Request - Guideline Checklist

Please check our [git workflow](https://github.com/DLR-SC/memilio/wiki/git-workflow). Use the **draft** feature if the Pull Request is not yet ready to review.

### Checks by code author

- [x] Every addressed issue is linked (use the "Closes #ISSUE" keyword below)
- [x] New code adheres to [coding guidelines](https://github.com/DLR-SC/memilio/wiki/Coding-guidelines)
- [x] No large data files have been added (files should in sum not exceed 100 KB, avoid PDFs, Word docs, etc.)
- [x] Tests are added for new functionality and a local test run was successful
- [x] Appropriate **documentation** for new functionality has been added (Doxygen in the code and Markdown files if necessary)
- [x] Proper attention to licenses, especially no new third-party software with conflicting license has been added

### Checks by code reviewer(s)

- [x] Corresponding issue(s) is/are linked and addressed
- [x] Code is clean of development artifacts (no deactivated or commented code lines, no debugging printouts, etc.)
- [x] Appropriate **unit tests** have been added, CI passes and code coverage is acceptable (did not decrease)
- [x] No large data files added in the whole history of commits(files should in sum not exceed 100 KB, avoid PDFs, Word docs, etc.)

### CODE REVIEW

**Comment by dabele** [comment]
feel free to add a reviewer if you feel someone else can better judge this.

**Review by xsaschako — COMMENTED** [review]
This looks good from my side, as it went through most of the ABM.
I went roughly through all things it should cover, and it looks good, here are my thoughts on some things:

- Ensemble_params_percentile and household probably shouldn't be included as they are setup/analysis
- infection/infection states are covered through randomization
- lockdown_rules are not included, I don't know if dampings make a difference, but could be
- locations can have npi's active, e.g. masks, for that to be tested it needs to be activated, that could lead to performance decreases
- trips are not measured

Thanks again for implementing, looking forward to the results, i will also tag @DavidKerkmann can have a second thought about what might be missing. I texted him and requested him.

**Comment by dabele** [comment]
>- lockdown_rules are not included, I don't know if dampings make a difference, but could be
>- locations can have npi's active, e.g. masks, for that to be tested it needs to be activated, that could lead to performance decreases
>- trips are not measured

I wouldn't expect adding dampings to make a difference, unless you are adding dampings dynamically during the simulation. I'll check.

Masks probably should be included.

Trips are not included intentionally, I feel they aren't really used or developed right now. They also aren't parallelized yet. So I didn't add that. Also, for trips we need a second benchmark anyway, I don't think this works in the same benchmark. We can add that later.

**Review by DavidKerkmann — APPROVED** [review]
Thanks for making this! I agree with Saschas comments for the most part. However, I could imagine having a separate benchmark for the pre- and post-processing as well, i.e., setup of the model and analysing the results.

**Comment by dabele** [comment]
Created Issue #835 to add the benchmark to CI

File: cpp/benchmarks/abm.cpp [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+
+    for (auto&& _ : state) {
+        state.PauseTiming(); //exclude the setup from the benchmark
+        auto sim = make_simulation(num_persons, seeds);
+        state.ResumeTiming();
+        sim.advance(sim.get_time() + mio::abm::days(10));
+    }
+}
+
+//Measure ABM simulation run time with different sizes and different seeds.
```
[the remark this answers is not in the document]
Comment (dabele) [reply cpp/benchmarks/abm.cpp:145]: added a comment why 10 days

File: cpp/benchmarks/abm.cpp [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+        home_size++;
+    }
+
+    //create other locations
+    for (auto loc_type : mio::enum_members<mio::abm::LocationType>()) {
+        if (loc_type == mio::abm::LocationType::Home) {
+            continue; //homes already created
+        }
+
+        const auto num_locs = std::max(size_t(1), num_persons / 5'000);
```
[the remark this answers is not in the document]
Comment (dabele) [reply cpp/benchmarks/abm.cpp]: I reduced the number a little bit. But keep in mind that many people who are assigned to schools don't go to school (i.e. adults), same for other places.

File: cpp/benchmarks/abm.cpp [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+                mio::UniformIntDistribution<size_t>::get_instance()(world.get_rng(), size_t(0), num_locs - 1);
+            person.set_assigned_location(locs[loc_idx]);
+        }
+    }
+
+    //infections and masks
+    for (auto& person : world.get_persons()) {
+        auto prng = mio::abm::Person::RandomNumberGenerator(world.get_rng(), person);
+        //~1% of people are infected, large enough to have some infection activity without everyone dying
+        if (mio::UniformDistribution<double>::get_instance()(prng, 0.0, 1.0) < 0.05) {
```
Comment (DavidKerkmann) [inline cpp/benchmarks/abm.cpp:54]: You say about 1% of people are infected, but then number is 0.05. Which one did you want to use?

**Review by DavidKerkmann — APPROVED** [review]
Thanks, this looks good. Only one small questions about the numbers.

File: cpp/benchmarks/abm.cpp (hunk 3 above)
Comment (dabele) [reply cpp/benchmarks/abm.cpp]: that's why you don't repeat a value in the comments...

I removed the number from the comment and introduced a variable that contains the percentage instead. Also did to some other places in the file.

**Review by DavidKerkmann — APPROVED** [review]

**Pull request merged**